zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java

```diff
@@ -156,10 +158,18 @@
      */
     private final Set<String> containers = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
 
+    private final Map<String, TTLNode> ttlMap = new HashMap<>();
```

Comment:
This addition of this map and the change of `ttls` from a concurrent map to a non-concurrent TreeSet breaks thread safety. The synchronization in DataTree is based on individual nodes yet these maps/sets belong to the DataTree itself. This is a serious latent bug and must be corrected or the implementation must be re-considered.

Reply:
I wonder if `ConcurrentSkipListMap` can be used (I'm not very familiar with it though).